// queryloader2/dist/queryloader2.js
(function e(t,n,r){function s(o,u){if(!n[o]){if(!t[o]){var a=typeof require=="function"&&require;if(!u&&a)return a(o,!0);if(i)return i(o,!0);var f=new Error("Cannot find module '"+o+"'");throw f.code="MODULE_NOT_FOUND",f}var l=n[o]={exports:{}};t[o][0].call(l.exports,function(e){var n=t[o][1][e];return s(n?n:e)},l,l.exports,e,t,n,r)}return n[o].exports}var i=typeof require=="function"&&require;for(var o=0;o<r.length;o++)s(r[o]);return s})({1:[function(require,module,exports){
'use strict';
var ImageLoaded = require('./ImageLoaded.js');

var QueryLoaderImage = {
  create: function() {
    this.element = document.createElement('img');
    this.element.setAttribute('src', this.src);
  },

  preload: function(cb) {
    ImageLoaded(this.element, function(err, alreadyLoaded) {
      cb(err, alreadyLoaded);
    });
  },
};

module.exports = function(src) {
  var image = Object.create(QueryLoaderImage);

  image.src = src;
  image.element = null;

  if (typeof src !== 'undefined') {
    image.create();
  }

  return image;
};

},{"./ImageLoaded.js":2}],2:[function(require,module,exports){
/*
 * Modified version of http://github.com/desandro/imagesloaded v2.1.1
 * MIT License. by Paul Irish et al.
 */

'use strict';
var BLANK = 'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///ywAAAAAAQABAAACAUwAOw==';

function loaded(image, callback) {
  var src;

  if (!image.nodeName) {
    return callback(new Error('First argument must be an image element'));
  }

  if (image.nodeName.toLowerCase() !== 'img') {
    return callback(new Error('Element supplied is not an image'));
  }

  if (image.src && image.complete && image.naturalWidth !== undefined) {
    return callback(null, true);
  }

  function bindEvent(element, event, attach, cb) {
    if (!element.addEventListener) {
      element[(attach ? 'attachEvent' : 'detachEvent')]('on' + event, cb);
    } else {
      element[(attach ? 'addEventListener' : 'removeEventListener')](event, cb);
    }
  }

  function onload() {
    bindEvent(image, 'load', false, onload);
    bindEvent(image, 'error', false, onload);

    callback(null, false);
  }

  bindEvent(image, 'load', true, onload);
  bindEvent(image, 'error', true, onload);

  if (image.readyState || image.complete) {
    src = image.src;
    image.src = BLANK;
    image.src = src;
  }
}

module.exports = loaded;

},{}],3:[function(require,module,exports){
'use strict';
var QueryLoaderImage = require('./Image.js');

var ImagePreloader = {
  getImageSrcs: function(element) {
    this.sources = [];

    if (typeof element !== 'undefined') {
      this.findImageInElement(element);

      if (this.deepSearch === true) {
        var elements = element.querySelectorAll('*');
        for (var i = 0; i < elements.length; i++) {
          if (elements[i].tagName !== 'SCRIPT') {
            this.findImageInElement(elements[i]);
          }
        }
      }
    }

    return this.sources;
  },

  findAndPreload: function(element) {
    if (typeof element === 'undefined') {
      return;
    }

    this.sources = this.getImageSrcs(element);

    for (var i = 0; i < this.sources.length; i++) {
      var image = QueryLoaderImage(this.sources[i]);
      image.preload(this.imageLoaded.bind(this));
      this.images.push(image);
    }
  },

  imageLoaded: function() {
    this.loaded++;

    this.updateProgress();
  },

  updateProgress: function() {
    this.parent.updateProgress(this.loaded, this.sources.length);
  },

  findImageInElement: function(element) {
    var urlType = this.determineUrlAndType(element);

    //skip if gradient
    if (!this.hasGradient(urlType.url)) {
      //remove unwanted chars
      urlType.url = this.stripUrl(urlType.url);

      //split urls
      var urls = urlType.url.split(', ');

      for (var i = 0; i < urls.length; i++) {
        if (this.validUrl(urls[i]) && this.urlIsNew(urls[i])) {
          var extra = '';

          if (this.isIE() || this.isOpera()) {
            //filthy always no cache for IE, sorry peeps!
            extra = '?rand=' + Math.random();
          }

          //add image to found list
          this.sources.push(urls[i] + extra);
        }
      }
    }
  },

  determineUrlAndType: function(element) {
    var url = '';
    var type = 'normal';
    var style = element.currentStyle || window.getComputedStyle(element, null);

    if ((typeof style.backgroundImage !== 'undefined' && style.backgroundImage !== '' && style.backgroundImage !== 'none')
      || (typeof element.style.backgroundImage !== 'undefined' && element.style.backgroundImage !== '' && element.style.backgroundImage !== 'none')
    ) {
      //if object has background image
      url = (style.backgroundImage || element.style.backgroundImage);
      type = 'background';
    } else if (typeof element.getAttribute('src') !== 'undefined' && element.nodeName.toLowerCase() === 'img') {
      //if is img and has src
      url = element.getAttribute('src');
    }

    return {
      url: url,
      type: type,
    };
  },

  hasGradient: function(url) {
    return (url && typeof url.indexOf !== 'undefined' ? url.indexOf('gradient(') !== -1 : false);
  },

  stripUrl: function(url) {
    url = url.replace(/url\(\'/g, '');
    url = url.replace(/url\(/g, '');
    url = url.replace(/\'\)/g, '');
    url = url.replace(/\)/g, '');

    return url;
  },

  validUrl: function(url) {
    if (url.length > 0 && !url.match(/^(data:)/i)) {
      return true;
    } else {
      return false;
    }
  },

  urlIsNew: function(url) {
    return this.sources.indexOf(url) === -1;
  },

  isIE: function() {
    return navigator.userAgent.match(/msie/i);
  },

  isOpera: function() {
    return navigator.userAgent.match(/Opera/i);
  },
};

module.exports = function(parent) {
  var imagePreloader = Object.create(ImagePreloader);

  imagePreloader.parent = parent;
  imagePreloader.sources = [];
  imagePreloader.images = [];
  imagePreloader.loaded = 0;
  imagePreloader.deepSearch = true;

  return imagePreloader;
};

},{"./Image.js":1}],4:[function(require,module,exports){
'use strict';
var PercentageParser = require('./PercentageParser');

var LoadingBar = {
  /**
   * Creates the element for the loading bar
   */
  create: function() {
    this.element = document.createElement('div');
    this.element.setAttribute('class', this.className);
    this.setStyling();
    this.updateProgress(0, 0);
  },

  setStyling: function() {
    this.element.style.height = this.barHeight + 'px';
    this.element.style.marginTop = '-' + (this.barHeight / 2) + 'px';
    this.element.style.backgroundColor = this.barColor;
    this.element.style.position = 'absolute';
    this.element.style.top = '50%';

    this.setTransitionTime(100);
  },

  updateProgress: function(percentage, time) {
    if (time !== 0) {
      this.setTransitionTime(time);
    }

    percentage = PercentageParser(percentage);

    this.element.style.width = percentage + '%';
  },

  setTransitionTime: function(ms) {
    this.element.style.WebkitTransition = 'width ' + ms + 'ms';
    this.element.style.MozTransition = 'width ' + ms + 'ms';
    this.element.style.OTransition = 'width ' + ms + 'ms';
    this.element.style.MsTransition = 'width ' + ms + 'ms';
    this.element.style.Transition = 'width ' + ms + 'ms';
  },
};

module.exports = function() {
  var loadingBar = Object.create(LoadingBar);

  loadingBar.element = null;
  loadingBar.className = 'queryloader__overlay__bar';
  loadingBar.barHeight = 1;
  loadingBar.barColor = '#fff';

  return loadingBar;
};

},{"./PercentageParser":6}],5:[function(require,module,exports){
'use strict';
var PercentageParser = require('./PercentageParser');

var Percentage = {
  create: function() {
    this.element = document.createElement('div');
    this.element.setAttribute('class', this.className);
    this.element.setAttribute('id', this.idName);
    this.applyStyling();
    this.updateProgress(0, 0);
  },

  applyStyling: function() {
    this.element.style.height = '40px';
    this.element.style.width = '100%';
    this.element.style.position = 'absolute';
    this.element.style.fontSize = '3em';
    this.element.style.top = '50%';
    this.element.style.left = '0';
    this.element.style.marginTop = '-' + (59 + this.barHeight) + 'px';
    this.element.style.textAlign = 'center';
    this.element.style.color = this.barColor;
  },

  updateProgress: function(percentage, time) {
    percentage = PercentageParser(percentage);

    this.element.innerHTML = percentage + '%';
  },
};

module.exports = function() {
  var percentage = Object.create(Percentage);

  percentage.element = null;
  percentage.idName = 'qlPercentage';
  percentage.className = 'queryloader__overlay__percentage';
  percentage.barHeight = 1;
  percentage.barColor = '#fff';

  return percentage;
};

},{"./PercentageParser":6}],6:[function(require,module,exports){
'use strict';
function parsePercentage(percentage) {
  if (parseInt(percentage) < 0) {
    return 0;
  }

  if (parseInt(percentage) > 100) {
    return 100;
  }

  return parseInt(percentage);
}

module.exports = parsePercentage;

},{}],7:[function(require,module,exports){
'use strict';
var LoadingBar = require('./LoadingBar');
var Percentage = require('./Percentage');

var Overlay = {
  init: function() {
    this.create();

    this.loadingBar = LoadingBar();
    this.loadingBar.barHeight = this.barHeight;
    this.loadingBar.barColor = this.barColor;
    this.loadingBar.create();
    this.element.appendChild(this.loadingBar.element);

    if (this.showPercentage) {
      this.percentage = Percentage();
      this.percentage.barColor = this.barColor;
      this.percentage.idName = this.percentageId;
      this.percentage.create();
      this.element.appendChild(this.percentage.element);
    }

    this.parentElement.appendChild(this.element);
  },

  create: function() {
    this.element = (document.querySelector('#' + this.idName) || document.createElement('div'));
    this.element.setAttribute('class', this.className);
    this.element.setAttribute('id', this.idName);
    this.applyStyling();
  },

  applyStyling: function() {
    //determine postion of overlay and set parent position
    this.element.style.position = this.calculatePosition();
    this.element.style.width = '100%';
    this.element.style.height = '100%';
    this.element.style.backgroundColor = this.backgroundColor;
    this.element.style.backgroundPosition = 'fixed';
    this.element.style.zIndex = 666999; //very HIGH
    this.element.style.top = '0';
    this.element.style.left = '0';

    this.element.style.WebkitTransition = 'opacity ' + this.fadeOutTime + 'ms';
    this.element.style.MozTransition = 'opacity ' + this.fadeOutTime + 'ms';
    this.element.style.OTransition = 'opacity ' + this.fadeOutTime + 'ms';
    this.element.style.MsTransition = 'opacity ' + this.fadeOutTime + 'ms';
    this.element.style.Transition = 'opacity ' + this.fadeOutTime + 'ms';
  },

  calculatePosition: function() {
    var overlayPosition = 'absolute';

    if (this.parentElement.tagName.toLowerCase() === 'body') {
      overlayPosition = 'fixed';
    } else {
      if (this.parentElement.style.position !== 'fixed' || this.parentElement.style.position !== 'absolute') {
        this.parentElement.style.position = 'relative';
      }
    }

    return overlayPosition;
  },

  updateProgress: function(percentage, time) {
    if (this.loadingBar !== null) {
      this.loadingBar.updateProgress(percentage, time);
    }

    if (this.percentage !== null) {
      this.percentage.updateProgress(percentage, time);
    }
  },

  remove: function() {
    if (this.canRemove(this.element)) {
      this.element.parentNode.removeChild(this.element);
    }
  },

  canRemove: function(element) {
    return (element.parentNode && typeof element.parentNode.removeChild !== 'undefined');
  },
};

module.exports = function(parentElement) {
  var overlay = Object.create(Overlay);

  overlay.parentElement = parentElement;
  overlay.idName = 'qLoverlay';
  overlay.percentageId = 'qlPercentage';
  overlay.className = 'queryloader__overlay';
  overlay.element = null;
  overlay.loadingBar = null;
  overlay.percentage = null;
  overlay.barColor = '#ff0000';
  overlay.backgroundColor = '#000';
  overlay.barHeight = 1;
  overlay.fadeOutTime = 300;
  overlay.showPercentage = false;

  return overlay;
};

},{"./LoadingBar":4,"./Percentage":5}],8:[function(require,module,exports){
if (!Function.prototype.bind) {
  Function.prototype.bind = function(oThis) {
    'use strict';
    if (typeof this !== 'function') {
      // closest thing possible to the ECMAScript 5
      // internal IsCallable function
      throw new TypeError('Function.prototype.bind - what is trying to be bound is not callable');
    }

    var aArgs = Array.prototype.slice.call(arguments, 1);
    var _this = this;
    var FNOP = function() {};

    var fBound = function() {
        return _this.apply(this instanceof FNOP && oThis
            ? this
            : oThis,
          aArgs.concat(Array.prototype.slice.call(arguments)));
      };

    FNOP.prototype = this.prototype;
    fBound.prototype = new FNOP();

    return fBound;
  };
}

if (typeof Object.create !== 'function') {
  // Production steps of ECMA-262, Edition 5, 15.2.3.5
  // Reference: http://es5.github.io/#x15.2.3.5
  Object.create = (function() {
    // To save on memory, use a shared constructor
    function Temp() {}

    // make a safe reference to Object.prototype.hasOwnProperty
    var hasOwn = Object.prototype.hasOwnProperty;

    return function(O) {
      // 1. If Type(O) is not Object or Null throw a TypeError exception.
      if (typeof O !== 'object') {
        throw TypeError('Object prototype may only be an Object or null');
      }

      // 2. Let obj be the result of creating a new object as if by the
      //    expression new Object() where Object is the standard built-in
      //    constructor with that name
      // 3. Set the [[Prototype]] internal property of obj to O.
      Temp.prototype = O;
      var obj = new Temp();
      Temp.prototype = null; // Let's not keep a stray reference to O...

      // 4. If the argument Properties is present and not undefined, add
      //    own properties to obj as if by calling the standard built-in
      //    function Object.defineProperties with arguments obj and
      //    Properties.
      if (arguments.length > 1) {
        // Object.defineProperties does ToObject on its first argument.
        var Properties = Object(arguments[1]);
        for (var prop in Properties) {
          if (hasOwn.call(Properties, prop)) {
            obj[prop] = Properties[prop];
          }
        }
      }

      // 5. Return obj
      return obj;
    };
  })();
}

},{}],9:[function(require,module,exports){
'use strict';
module.exports = function() {
  var timeout = window.setInterval(function() {
    if (typeof document.getElementsByTagName('body')[0] !== 'undefined') {
      var tempOverlay = document.createElement('div');
      tempOverlay.style.position = 'fixed';
      tempOverlay.style.width = '100%';
      tempOverlay.style.height = '100%';
      tempOverlay.style.zIndex = '9999';
      tempOverlay.style.backgroundColor = '#000';
      tempOverlay.style.left = '0';
      tempOverlay.style.top = '0';
      tempOverlay.setAttribute('id', 'qLtempOverlay');
      document.getElementsByTagName('body')[0].appendChild(tempOverlay);

      window.clearInterval(timeout);
    }
  }, 1);
};

},{}],10:[function(require,module,exports){
'use strict';
var ImagePreloader = require('./../ImagePreloader');
var Overlay = require('./../Overlay');

var QueryLoader = {
  init: function(options, defaultOptions) {
    this.options = this.extend(defaultOptions, options);

    if (typeof this.element !== 'undefined') {
      this.createOverlay();
      this.removeTempOverlay();
      this.createPreloader();
      this.startMaxTimeout();
    }
  },

  extend: function(base, adding) {
    if (typeof base === 'undefined') {
      base = {};
    }

    for (var property in adding) {
      if (adding.hasOwnProperty(property)) {
        base[property] = adding[property];
      }
    }

    return base;
  },

  startMaxTimeout: function() {
    this.maxTimeout = window.setTimeout(this.doneLoading.bind(this), this.options.maxTime);
  },

  createOverlay: function() {
    this.overlay = Overlay(this.element);
    this.overlay.idName = this.options.overlayId;
    this.overlay.percentageId = this.options.percentageId;
    this.overlay.backgroundColor = this.options.backgroundColor;
    this.overlay.barHeight = this.options.barHeight;
    this.overlay.barColor = this.options.barColor;
    this.overlay.showPercentage = this.options.percentage;
    this.overlay.fadeOutTime = this.options.fadeOutTime;

    if (typeof this.element !== 'undefined') {
      this.overlay.init();
    }
  },

  removeTempOverlay: function() {
    window.setTimeout(function() {
      var tempOverlay = document.getElementById('qLtempOverlay');
      if (tempOverlay && tempOverlay.parentNode) {
        tempOverlay.parentNode.removeChild(tempOverlay);
      }
    }, 0);
  },

  createPreloader: function() {
    this.preloader = ImagePreloader(this);
    this.preloader.deepSearch = this.options.deepSearch;

    window.setTimeout(function() {
      this.preloader.findAndPreload(this.element);
    }.bind(this), 100);
  },

  updateProgress: function(done, total) {
    var percentage = ((done / total) * 100);
    this.overlay.updateProgress(percentage, this.options.minimumTime);

    if (typeof this.options.onProgress === 'function') {
      this.options.onProgress(percentage, done, total);
    }

    if (done === total && this.done === false) {
      window.clearTimeout(this.maxTimeout);
      window.setTimeout(this.doneLoading.bind(this), this.options.minimumTime);
    }
  },

  doneLoading: function() {
    window.clearTimeout(this.maxTimeout);
    this.done = true;

    this.overlay.element.style.opacity = 0;

    window.setTimeout(this.destroy.bind(this), this.options.fadeOutTime);
  },

  destroy: function() {
    this.overlay.remove();

    this.options.onComplete();
  },
};

module.exports = function(element, options) {
  var queryLoader = Object.create(QueryLoader);

  queryLoader.element = element;
  queryLoader.options = {};
  queryLoader.done = false;
  queryLoader.maxTimeout = null;

  //children
  queryLoader.overlay = null;
  queryLoader.preloader = null;

  if (element !== null) {
    var voidFunc = function() {};

    queryLoader.init(options, {
      onComplete: voidFunc,
      onProgress: voidFunc,
      backgroundColor: '#000',
      barColor: '#fff',
      overlayId: 'qLoverlay',
      percentageId: 'qLpercentage',
      barHeight: 1,
      percentage: false,
      deepSearch: true,
      minimumTime: 300,
      maxTime: 10000,
      fadeOutTime: 1000,
    });
  }

  return queryLoader;
};

},{"./../ImagePreloader":3,"./../Overlay":7}],11:[function(require,module,exports){
require('./Polyfills');
var QueryLoader2 = require('./QueryLoader');
var TempOverlay = require('./QueryLoader/TempOverlay');

//jquery and zepto
if (window.jQuery || window.Zepto) {
  (function($) {
    'use strict';
    $.fn.queryLoader2 = function(options) {
      return this.each(function() {
        (QueryLoader2(this, options));
      });
    };
  })(window.jQuery || window.Zepto);
}

// component
if (typeof module !== 'undefined') {
  module.exports = QueryLoader2;
}

// requirejs support
if (typeof define === 'function' && define.amd) {
  define([], function() {
    'use strict';
    return QueryLoader2;
  });
}

window.QueryLoader2 = QueryLoader2;
TempOverlay();

},{"./Polyfills":8,"./QueryLoader":10,"./QueryLoader/TempOverlay":9}]},{},[11]);
